### PR TITLE
Preview Markdown while commenting on a submission

### PR DESCRIPTION
**Pull request by ktuite**
The comments in the Submission activity feed support Markdown, and now that Markdown can be previewed as it is typed!

Screen Shot 2021-08-06 at 2 40 34 PM

There is a new reusable component called `MarkdownTextarea` that is used in the two places we accept comments on Submissions:
- Submissions table when changing the review state
- Submission activity feed

There are a few configurable properties on this `MarkdownTextarea`:
1. The prop in the parent component that will be bound to a prop called `value`. This is so the outer component can continue to handle the comment submission/POST request.
2. The parent compnent's `awaitingRepsonse` prop so some of the UI elements will still be displayed during requests.
3. If the footer with the link "Markdown supported" and "Preview" button should be shown when the component is initialized, or if it should wait until the user starts typing text.
4. Default placerholder text
5. Additional `<slot>` HTML for any buttons that appear to the right of the "Preview" button. 

### Markdown Preview on Submissions table comments
(Comment submit button is the "Update" button on bottom left, not next to the "Preview" button)
Screen Shot 2021-08-06 at 2 43 33 PM
Screen Shot 2021-08-06 at 2 43 54 PM
Screen Shot 2021-08-06 at 2 43 58 PM

### Markdown Preview on Submission activity feed when Comment for an Edit is needed
(Shows footer and active blue border by default)
Screen Shot 2021-08-06 at 2 38 17 PM
Screen Shot 2021-08-06 at 2 39 05 PM
Screen Shot 2021-08-06 at 2 39 09 PM

### Markdown Preview on Submission activity feed otherwise
(Doesn't show the footer by default)
Screen Shot 2021-08-06 at 2 59 48 PM
Screen Shot 2021-08-06 at 2 59 57 PM
Screen Shot 2021-08-06 at 3 00 22 PM
Screen Shot 2021-08-06 at 3 00 28 PM

### CODE REVIEW

File: src/components/submission/markdown-textarea.vue [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  "en": {
+    // This is a link to an external website with Markdown style guidelines
+    "markdownSupported": "Markdown supported",
+    "preview": "Preview",
+    "hidePreview": "Hide Preview"
+  }
+}
+</i18n>
+
+<!-- Autogenerated by destructure.js -->
```
Comment (matthew-white) [inline src/components/submission/markdown-textarea.vue:174]: OK if this isn't moved here.

File: src/components/submission/comment.vue [hunk 2]
```diff
@@ -164,28 +138,22 @@ export default {
 <i18n>
 {
   "cs": {
-    "editWithoutComment": "Provedli jste úpravy těchto údajů. Popište prosím provedené změny.",
-    "markdownSupported": "Podporován formát Markdown"
```
Comment (matthew-white) [inline src/components/submission/comment.vue:168]: OK if this isn't removed.

File: src/components/submission/markdown-textarea.vue [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+<template>
+  <div id="markdown-textarea" :class="activeBorderClass">
+    <div class="form-group">
+      <textarea :value="value" class="form-control"
+        :placeholder="defaultText" :aria-label="defaultText"
+        required rows="2" @input="update">
+      </textarea>
+    </div>
+    <div v-if="value !== '' && previewMode" id="preview-container">
+      <p class="heading">Preview</p>
```
Comment (matthew-white) [inline src/components/submission/markdown-textarea.vue:21]: i18n message for "Preview"

File: src/components/submission/markdown-textarea.vue [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+https://github.com/getodk/central-frontend/blob/master/NOTICE.
+
+This file is part of ODK Central. It is subject to the license terms in
+the LICENSE file found in the top-level directory of this distribution and at
+https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+including this file, may be copied, modified, propagated, or distributed
+except according to the terms contained in the LICENSE file.
+-->
+<template>
+  <div id="markdown-textarea" :class="activeBorderClass">
```
Comment (matthew-white) [inline src/components/submission/markdown-textarea.vue:13]: id -> class

File: src/components/submission/markdown-textarea.vue [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    <div class="form-group">
+      <textarea :value="value" class="form-control"
+        :placeholder="defaultText" :aria-label="defaultText"
+        required rows="2" @input="update">
+      </textarea>
+    </div>
+    <div v-if="value !== '' && previewMode" id="preview-container">
```
Comment (matthew-white) [inline src/components/submission/markdown-textarea.vue:20]: id -> class

File: src/components/submission/markdown-textarea.vue [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+      <p class="heading">Preview</p>
+      <!-- eslint-disable-next-line vue/no-v-html -->
+      <div v-html="valueMarkdown"></div>
+    </div>
+    <div v-show="value !== '' || showFooter || parentAwaitingResponse" id="submission-comment-actions">
```
Comment (matthew-white) [inline src/components/submission/markdown-textarea.vue:25]: id -> class (and maybe change the class name?)

File: src/components/submission/markdown-textarea.vue [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+      <a href="https://commonmark.org/help/" class="external-help-link" target="_blank" rel="noopener">{{ $t('markdownSupported') }} </a>
+      <span class="push"></span>
+      <button class="btn md-preview-btn" @click.prevent="toggleViewMode">
```
Comment (matthew-white) [inline src/components/submission/markdown-textarea.vue:28]: `.prevent` might not be needed if you use `type="button"`.

**Review by matthew-white — COMMENTED** [review]
Notes from interactive review

File: src/components/markdown/textarea.vue [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+</script>
+
+<style lang="scss">
+
+.markdown-textarea {
+  &.active-border {
+    border: 1px solid #31708f;
+  }
+
+  background: #eaeaea;
```
[the remark this answers is not in the document]
Comment (ktuite) [reply src/components/markdown/textarea.vue:86]: I wanted the background of this component to be slightly darker than the text area above it but i will use `darken( $color-subpanel-background, 3% );` to accomplish that

File: src/components/markdown/textarea.vue [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+<i18n lang="json5">
+{
+  "en": {
+    // This is a link to an external website with Markdown style guidelines
+    "markdownSupported": "Markdown supported",
+    // This is text shown as a label above the preview of Markdown formatting (special formatting of user-submitted text)
+    "preview": "Preview",
+    // This text is shown on a button that the user clicks to change if the Markdown preview is shown or not
+    "toggle": {
+      "showPreview": "Preview",
```
[the remark this answers is not in the document]
Comment (ktuite) [reply src/components/markdown/textarea.vue:146]: Sounds good - i'm going to put the hide preview there, too.

File: test/components/submission/comment.spec.js [hunk 10]
```diff
@@ -99,29 +100,17 @@ describe('SubmissionComment', () => {
...
-      const actions = component.get('#submission-comment-actions');
-      actions.should.be.hidden();
-      const textarea = component.get('textarea');
-      await textarea.setValue('foo');
-      actions.should.be.visible();
-      await textarea.setValue('');
-      actions.should.be.hidden();
-    });
-
+  describe('visibility of actions during requests', () => {
```
[the remark this answers is not in the document]
Comment (ktuite) [reply test/components/submission/comment.spec.js:103]: I had renamed this block to add "with requests" for some reason, but also moved one of the tests to `textarea.spec.js`, so the title doesn't really make sense. I am going to just remove this outer describe and keep the inner tests.

**Pull request merged**